### PR TITLE
replace types of Union with Union of types in some places

### PR DESCRIPTION
**Pull request by rfourquet**
`T <: Union{A, B}` is sometimes meant to mean either `A` or `B`, which are concrete types. This can result in crashes or wrong results if T takes actually the value `Union{A, B}`.

There may be other places in base with similar bugs...

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]
nice finds :)

**Comment by JeffBezanson** [comment]
Thanks for finding these. I would like to try to minimize use of `@eval` though; it makes the code harder to read and duplicates the method bodies. Types like `Union{Array{T1}, Array{T2}, ...}` can be used instead of `Array{T} where T<:Union{T1,T2}`.

**Comment by rfourquet** [comment]
> `@eval` [...] duplicates the method bodies.

OK, I didn't know there would be a difference for the compiler.

> Types like Union{Array{T1}, Array{T2}, ...} can be used instead of Array{T} where T<:Union{T1,T2}.

I started in a couple of places, but when the list of types is big, it start to be quite verbose... Please have a look at this patch in array.jl, and if it's fine by you, I will continue the transormation in the remaining places.

File: base/random.jl
Comment (tkelman) [inline base/random.jl]: indent off a bit

File: base/array.jl
Comment (JeffBezanson) [inline base/array.jl]: You can use something like `const BitIntegerArray = Union{map(T->Array{T}, uniontypes(BitInteger))...}`.

Comment (rfourquet) [inline base/array.jl]: Ok thanks, done.

File: base/array.jl
Comment (tkelman) [inline base/array.jl]: we don't use `A` for an array type in that many signatures, `T` or something a few characters longer would be a more common usage

File: base/random.jl
Comment (tkelman) [inline base/random.jl]: would be better to remove the blank line at the start of the `if is_windows()` block than add more at the beginning and end here - this is redundant with the indentation

File: base/random.jl
Comment (tkelman) [inline base/random.jl]: match

File: base/random.jl
Comment (rfourquet) [inline base/random.jl]: I usually prefer to have space before and after struct/method definitions for better readability (i.e. this if/else/end is not usual code), but your call.

File: base/array.jl
Comment (rfourquet) [inline base/array.jl]: I found it useful for distinction as a composite type, but ok.

Comment (rfourquet) [inline base/array.jl]: So `T` was looking like an`eltype`, so I chose `Arr`; if still not good, please suggest.

File: base/random.jl
Comment (tkelman) [inline base/random.jl]: what isn't usual code about it?

File: base/array.jl
Comment (tkelman) [inline base/array.jl]: Arr works for me, it's not a variable name that we heavily use for array values

File: base/random.jl
Comment (rfourquet) [inline base/random.jl]: I just meant it's rare to have struct or method definition within an if block, it's usually at the top level; sometimes the style is amended for special cases, like skipping the indentation corresponding module blocks, I thought why not here, but no pb!

File: base/array.jl
Comment (rfourquet) [inline base/array.jl]: (Quite an understatement as I could not find any use of `Arr` elsewhere in Base ;-) )

**Comment by rfourquet** [comment]
Good to go?

**Comment by rfourquet** [comment]
Will merge in a couple of days if no objection.

**Pull request merged**
